### PR TITLE
Allow ingesting from a non-default S3 endpoint

### LINKED ISSUE #2886: lakectl ingest from Minio
Is it possible to use `lakectl ingest ` to get data from minio instead of amazon s3 ?

### PR DESCRIPTION
**Pull request by tshradheya**
Fixes #2886 

Command used to ingest:

```
 LAKECTL_S3_ENDPOINT_URL="http://127.0.0.1:9000"
  ./lakectl ingest --from s3://src-buck --to lakefs://dest-buck/main/ --intg minio 

```

### CODE REVIEW

**Comment by tshradheya** [comment]
@johnnyaug i have coded a working feature but need some advice on why this happens when i ingest the data 

image

The data is always referenced inside an empty directory 

Also whenever i click to download the file, it download a `Failed-Server prolem objects.json`

**Comment by johnnyaug** [comment]
@tshradheya, thank you for tackling this issue!
1. Regarding the empty directory, I think it's a bug in the existing ingest code. I will look into it and open an issue.
1. About the error you're getting, maybe your installation cannot access your minio storage? Is lakeFS configured on top of minio? Are you able to upload a file into your repository from the UI?
1. In general, I think the implementation should not use a separate MinIO client. Instead, I would use the MinIO S3-compatible API. You can introduce a new LAKECTL_S3_ENDPOINT_URL variable, and use the same S3 client - only adding the endpoint parameter to use MinIO instead of S3.

**Comment by tshradheya** [comment]
1. Thanks, appreciate the help
2. Currently my local lakefs has a repository with namespace of a local minio server. And i am trying ingesting from http://play.min.io . From my understanding after ingestion I should be able to read that data from my local minio server? Please correct my understanding if i am wrong
3. Thanks for this suggestion! New to the world of S3 and block storage so let me read up on this and change implementation accordingly 😄

**Comment by johnnyaug** [comment]
I think this feature is a bit more tricky than we thought.
As mentioned in our [docs for ingest](https://docs.lakefs.io/setup/import.html#zero-copy-import), the lakeFS installation needs to be able to read the source objects. Note that ingesting does not copy any objects - it only creates pointers to the objects in the source bucket.

In the case of MinIO - it also means that the installation **needs to use the same MinIO installation as the source.** We will have to document this once this is merged.
So in short, you can't ingest data from one MinIO installation to another- you can only ingest from a MinIO bucket to a lakeFS repository which is on top of the same MinIO installation.

**Comment by tshradheya** [comment]
> In the case of MinIO - it also means that the installation needs to use the same MinIO installation as the source. We will have to document this once this is merged.

Yes, this makes more sense in my head now haha
I fixed my issue of download and it was because of wrong address. 


Will work on simplifying the logic as you said and then update the docs as well :)

**Comment by tshradheya** [comment]
> In general, I think the implementation should not use a separate MinIO client. Instead, I would use the MinIO S3-compatible API. You can introduce a new LAKECTL_S3_ENDPOINT_URL variable, and use the same S3 client - only adding the endpoint parameter to use MinIO instead of S3.

I have added three different variable for minio when ingesting. This will allow ingestion from Minio to any S3 blockstore like AWS S3 as well if my understanding is correct. 
I have updated the implementation, please do take a look and let me know if anything needs to be revised and then i can accordingly document

Ingest command:
```
MINIO_STORAGE_ACCOUNT="127.0.0.1:9000" MINIO_STORAGE_ACCESS_KEY="minioadmin" MINIO_STORAGE_SECRET_ACCESS_KEY="minioadmin"
  ./lakectl ingest --from s3://src-buck --to lakefs://dest-buck/main/ --intg minio

```

File: cmd/lakectl/cmd/ingest.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/ingest.go:136]:
```suggestion
	ingestCmd.Flags().String("integration-source", "", "integration source to use for ingestion")
```

"intg" is a big hard for me to understand.  I'd be happy for an alias or even a one-letter short flag if `integration-source` is too much to type.  But I _do_ like the descriptive flag name for a flag which is not used much, it is easier for me to read.

File: cmd/lakectl/cmd/store/s3.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/store/s3.go:20]: I cannot find documentation for these environment variables.  There is [no reference](https://www.google.com/search?q=site%3Amin.io+"MINIO_STORAGE_ACCOUNT") to these environment variables in min.io or on [the Minio organization on GitHub](https://github.com/search?q=org%3Aminio+%22MINIO_STORAGE_ACCOUNT%22&type=code).

**Review by arielshaqed — COMMENTED** [review]
Hi @tshradheya,

Thanks for this extension of `lakectl ingest` to work with MinIO!  As you can probably imagine, we _really_ want it to be easy to ingest data into lakeFS.

I am having difficulty understanding the solution in this PR, possibly because I come from a more AWS background.  It should of course be possible to hook up `lakectl` to any S3-like source.  But as an S3 user I would expect that to happen by configuring the regular `AWS_*` environment variables (either `AWS_PROFILE` or `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` and possibly a few more).  So what I would primarily be missing might be a way to specify a different S3 endpoint URL for lakectl to use.

I am not sure how many of our users will have these `MINIO_*` environment variables set.  Ideally these variables would be appear in minIO documentation (maybe for the `mc` command?).  Then the documentation for the `integration-source` flag that you are adding could link to that documentation.  Of course these variables might commonly be used in the MinIO community -- we could link to any page that talks about them, really.

Thanks!  Really looking forward to improving onboarding from MinIO to lakeFS.

**Comment by tshradheya** [comment]
Hi, thanks for the feedback.

You are right, these env variables aren't in anyway related to MINIO_*

I wanted to check if the following scenario is possible when lakefs has namespace from AWS S3 but wants to ingest data from MinIO?

To support the above, is the main reason why I introduced my own MINIO env variables so they don't clash with the S3 variables set by lakefs

If above doesn't need to supported then i can simplify the logic to just accept S3_ENDPOINT and use the already defined AWS tokens for the credentials.


Once the implementation has a firstpass i will write up documentation for this as well 😀

**Comment by arielshaqed** [comment]
> Hi, thanks for the feedback.

Sure, @tshradheya !  As you can imagine, it's really important for us to make it easy to ingest data into lakeFS.  So thanks for helping to do that!

> 
> You are right, these env variables aren't in anyway related to MINIO_*
> 
> I wanted to check if the following scenario is possible when lakefs has namespace from AWS S3 but wants to ingest data from MinIO?

This will be difficult, but for other reasons.  "Ingest" is not a copy operation, rather it creates metadata on the lakeFS side that points at existing objects.  So those objects must be accessible from the same S3 endpoint that lakeFS uses.

There is no support currently for _multiple block adapters_.  The challenge here is not so much to code it as to ensure consistent behaviour and semantics.  We'll need (several) very concrete use-cases for such support before we can implement it.

So you will need to copy your files from MinIO to lakeFS over S3.  There are numerous 3rd-party utilities to perform bulk S3 copies, depending on which tooling you are most familiar with and your performance requirements.  I would probably start with our integration guide [Copying data with Rclone](https://docs.lakefs.io/integrations/rclone.html), and just copy everything from MinIO into lakeFS and then commit.
If performance turns out to be a limiting factor then you have a _lot_ of data, so you will want to have a Spark cluster.  Assuming you do, you could [use DistCp to copy data](https://docs.lakefs.io/integrations/distcp.html).

> 
> To support the above, is the main reason why I introduced my own MINIO env variables so they don't clash with the S3 variables set by lakefs
> 
> If above doesn't need to supported then i can simplify the logic to just accept S3_ENDPOINT and use the already defined AWS tokens for the credentials.

Your code only touches `lakectl` (the command-line tool), and `lakefs` (the daemon/server) will be entirely unaffected by whatever settings you change here.  So there is no need to read credentials from another source, just use whatever credentials mechanism in boto is easiest for you.

You *will* most likely need to change the endpoint URL, so I would consider adding a CLI flag that changes that.
 
> Once the implementation has a firstpass i will write up documentation for this as well grinning

Sure!  Thanks for taking the time to explain.  Just please note you will *not* be able to ingest data from MinIO into a lakeFS that runs on S3, you will need to copy as explained above.

If you do need a lakeFS that can handle multiple backends at once, please open an issue.  It will be a **big** discussion, and we will have some questions.  🐘 ❓ 🐘 ❓ 🐘

**Comment by tshradheya** [comment]
@arielshaqed Thanks for being patient and explaining 😄 

I now have a much better understanding of how things work.

I have updated the implementation to use the shared config from AWS CLI. I am thinking we should support `config.Profile` as well since people may have multiple S3 creds stored in AWS CLI 

Updated the docs as well! Would be great if you could help take a look :)

File: cmd/lakectl/cmd/store/s3.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/store/s3.go:37]: 1. Can we get the URL from a flag, instead, say `s3-endpoint-url`?  Every other lakectl option uses flags, this should too.
1. I would rather not have to mention "minio" at all.  If I specify `--s3-endpoint-url`, then presumably I want to use that endpoint.  It might be MinIO, it might be any other S3 compatible storage, it might even be AWS S3 just using a different endpoint.  So I would drop the flag `integration-source` entirely.

File: cmd/lakectl/cmd/store/s3.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/store/s3.go:43]: nit:
```suggestion
```
We prefer not to accumulate blank lines at the ends of functions.

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Thanks!

I'd be happy to approve if we have just an `endpoint-url` flag, and no `integration-source` flag.

Please note that ingestion does not copy objects.  So this code _will_ ingest objects from your MinIO installation to a lakeFS set up to use that same MinIO installation.  I believe you indicated that your use-case might be to copy files on a MinIO installation into a lakeFS set up to write to S3.  This code will **not** be able to make that work.

File: cmd/lakectl/cmd/store/s3.go
Comment (tshradheya) [reply cmd/lakectl/cmd/store/s3.go:37]: Nice, that makes it much more clearer 👍

**Comment by tshradheya** [comment]
> I believe you indicated that your use-case might be to copy files on a MinIO installation into a lakeFS set up to write to S3. This code will not be able to make that work.

Yes it makes sense to me now. I currently dont need that functionality but was assuming that this might have to support that as well since both are S3 compatible but as you explained i was wrong to assume so

**Comment by tshradheya** [comment]
@arielshaqed could you help review the updated PR? Thanks :)

File: cmd/lakectl/cmd/ingest.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/ingest.go:136]:
```suggestion
	ingestCmd.Flags().String("s3-endpoint-url", "", "URL to access S3 storage API (by default, use regular AWS S3 endpoint)")
```

File: cmd/lakectl/cmd/store/s3.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/store/s3.go:18]:
```suggestion
	if s3EndpointUrl != "" {
```

File: cmd/lakectl/cmd/store/s3.go
Comment (arielshaqed) [inline cmd/lakectl/cmd/store/s3.go:31]: (Not required!)

Usually I prefer to extract shared code outside of conditionals.  Here this would be dropping the :
```suggestion
	var config aws.Config{}
	if s3EndpointUrl != "" {
		config = aws.Config{
			Endpoint:         aws.String(s3EndpointUrl),
			Region:           aws.String("us-east-1"), // Needs region for validation as it is AWS client
			S3ForcePathStyle: aws.Bool(true),
			}
	}
	sess, err := session.NewSessionWithOptions(session.Options{
		SharedConfigState: session.SharedConfigEnable,
		Config: config,
	})
	if err != nil {
		return nil, err
	}
```

File: docs/setup/import.md
Comment (arielshaqed) [inline docs/setup/import.md:51]: I would shorten phrasing here:
```suggestion
Specify an endpoint to ingest from other S3 compatible storage solutions, e.g. add `--s3-endpoint-url https://play.min.io`.
```

**Review by arielshaqed — COMMENTED** [review]
Thanks for your contribution and you patience with me!

I'm running the tests and everything should be cool now.  I would obviously prefer for you to handle the minor comments I added in this round, but that is **not** required.

I am renaming this PR to better reflect what it does.

**Review by arielshaqed — APPROVED** [review]
_I had **one job**_... and I forgot to approve.  Sorry :-/

**Comment by arielshaqed** [comment]
Please note test and lint failures.  No worries, this is just CI/CD doing its job.  Let me know if you need help fixing them.

**Comment by arielshaqed** [comment]
Thanks!  We should be good to go as soon as tests pass :-)

**Comment by arielshaqed** [comment]
Tests passed on #3140!

**Comment by arielshaqed** [comment]
enforce-label check is incorrect, everything else is green ✅ , so pulling with extreme administrative force.

Thanks again, @tshradheya !

**Pull request merged**
